### PR TITLE
Add U-Net model

### PR DESCRIPTION
**Pull request by adamjstewart**
This PR adds the U-Net model to `torchvision`. U-Net is very popular in image segmentation, especially in the biomedical imaging space.

Questions: 

- What's the requirement for pretrained models? I'm not sure what the standard dataset for image segmentation is, or which hyperparameters would work best. I just tried to follow the original U-Net paper as best I could.
- The unit tests seem to be failing. It looks like all segmentation models are supposed to be defined in `segmentation.py`? This seems cumbersome. Any suggestions for how to improve this?

### CODE REVIEW

**Comment by fmassa** [comment]
Thanks for the PR!

> What's the requirement for pretrained models? I'm not sure what the standard dataset for image segmentation is, or which hyperparameters would work best. I just tried to follow the original U-Net paper as best I could.

We are currently focusing on COCO / Pascal for semantic segmentation tasks. I'll be uploading pre-trained weights for those models soon, and they have been trained with the `references/segmentation/train.py` script available in torchvision
It is a requirement to have pre-trained weights, and they should match the reported accuracies within a few %.
I could try giving a shot and train the model, but I won't have time before July I think.

> The unit tests seem to be failing. It looks like all segmentation models are supposed to be defined in segmentation.py? This seems cumbersome. Any suggestions for how to improve this?

My thoughts on this is that we should create a folder `segmentation`, and move all semantic segmentation models there. This will require a bit of refactoring, but I'm considering doing it anyway following the detection PR that I'm preparing.

**Comment by fmassa** [comment]
Also, at least for Pascal / COCO, we generally allow the model to take arbitrary backbones, so that one can switch from resnet to resnext for example, and reuse pre-trained weights. It seems that your implementation here doesn't follow this pattern?

**Comment by adamjstewart** [comment]
> I could try giving a shot and train the model, but I won't have time before July I think.

That's fine by me, no rush from my end.

> My thoughts on this is that we should create a folder segmentation, and move all semantic segmentation models there. This will require a bit of refactoring, but I'm considering doing it anyway following the detection PR that I'm preparing.

I agree, I'm fine with waiting until you finish your detection PR. I think there should be a folder for classification, object_detection, semantic_segmentation, and instance_segmentation.

> Also, at least for Pascal / COCO, we generally allow the model to take arbitrary backbones, so that one can switch from resnet to resnext for example, and reuse pre-trained weights. It seems that your implementation here doesn't follow this pattern?

I'm not sure if U-Nets fit that description. ResNets usually condense an image down to a single pixel (with multiple channels), followed by fully connected layers. U-Nets condense an image to a smaller 32x32 grid, then upsample the image to its original resolution.

File: torchvision/models/unet.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+import torch.nn as nn
+
+
+__all__ = ['UNet', 'unet8', 'unet13', 'unet18', 'unet23', 'unet28', 'unet33']
+
+
+def double_conv(in_channels, out_channels):
+    return nn.Sequential(
+        nn.Conv2d(in_channels, out_channels, 3),
+        nn.ReLU(),
```
[the remark this answers is not in the document]
Comment (adamjstewart) [reply torchvision/models/unet.py]: Thanks! 

P.S. Nice Akame pic!

File: torchvision/models/unet.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    Args:
+        in_channels (int, optional): number of channels in input image
+        out_channels (int, optional): number of channels in output segmentation
+        start_channels (int, optional): power of 2 channels to start with
+        depth (int, optional): number of contractions/expansions
+        p (float, optional): dropout probability
+    """
+
+    def __init__(self, in_channels=1, out_channels=2, start_channels=6,
```
Comment (ekagra-ranjan) [inline torchvision/models/unet.py]: Does `out_channels` here correspond to the number of heatmaps, each representing a heatmap for an individual class?If yes then you might consider renaming it to `num_classes` following the naming convention of pytorch models.

Comment (adamjstewart) [reply torchvision/models/unet.py]: Yes, that's what it represents. I went with `out_channels` because there didn't seem to be a consensus among existing models:
```console
$ grep '__init__.*num_classes' *
alexnet.py:    def __init__(self, num_classes=1000):
deeplabv3.py:    def __init__(self, in_channels, num_classes):
googlenet.py:    def __init__(self, num_classes=1000, aux_logits=True, transform_input=False, init_weights=True):
googlenet.py:    def __init__(self, in_channels, num_classes):
inception.py:    def __init__(self, num_classes=1000, aux_logits=True, transform_input=False):
inception.py:    def __init__(self, in_channels, num_classes):
mobilenet.py:    def __init__(self, num_classes=1000, width_mult=1.0):
resnet.py:    def __init__(self, block, layers, num_classes=1000, zero_init_residual=False,
shufflenetv2.py:    def __init__(self, stages_repeats, stages_out_channels, num_classes=1000):
squeezenet.py:    def __init__(self, version='1_0', num_classes=1000):
vgg.py:    def __init__(self, features, num_classes=1000, init_weights=True):
$ grep '__init__.*out_channels' *
deeplabv3.py:    def __init__(self, in_channels, out_channels, dilation):
deeplabv3.py:    def __init__(self, in_channels, out_channels):
googlenet.py:    def __init__(self, in_channels, out_channels, **kwargs):
inception.py:    def __init__(self, in_channels, out_channels, **kwargs):
shufflenetv2.py:    def __init__(self, stages_repeats, stages_out_channels, num_classes=1000):
unet.py:    def __init__(self, in_channels, out_channels, dropout=False, p=0.5):
unet.py:    def __init__(self, in_channels, out_channels):
unet.py:    def __init__(self, in_channels=1, out_channels=2, start_channels=6,
```
I'll switch to `num_classes`.

Comment (adamjstewart) [reply torchvision/models/unet.py]: Okay, switched it to `num_classes`. The helper functions still use `out_channels` since that number is independent of the final number of classes at the end.

File: torchvision/models/unet.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+        x = self.conv(x)
+
+        return x
+
+
+class UNet(nn.Module):
+    """`U-Net <https://arxiv.org/pdf/1505.04597.pdf>`_ architecture.
+
+    Args:
+        in_channels (int, optional): number of channels in input image
```
Comment (adamjstewart) [inline torchvision/models/unet.py]: So the reason I chose `in_channels=1` as the default for U-Net is because this is how the original U-Net paper is modeled, using a single channel grayscale microscope imagery dataset (see #900). The application I needed it for was actually 4-channel microscope imagery, but unfortunately PIL doesn't support this (see #882). If we decide to pretrain this on COCO/Pascal I'm fine with switching the default to 3-channel.

**Comment by rwightman** [comment]
> > Also, at least for Pascal / COCO, we generally allow the model to take arbitrary backbones, so that one can switch from resnet to resnext for example, and reuse pre-trained weights. It seems that your implementation here doesn't follow this pattern?
> 
> I'm not sure if U-Nets fit that description. ResNets usually condense an image down to a single pixel (with multiple channels), followed by fully connected layers. U-Nets condense an image to a smaller 32x32 grid, then upsample the image to its original resolution.

It's pretty common for the encoder half of the network to be based on a standard backbone like ResNet or VGG. If I recall, the original U-Net paper was pretty a much a VGG net on the encoder side already.

For the decoder, many find that using upsampling provides better results than transpose convolutions. I've seen a few impl that allow choosing either when the model is constructed.

**Comment by rwightman** [comment]
Further my previous comment, some PyTorch U-Nets with support for different backbones and the mentioned upsampling decoder blocks.

* https://github.com/qubvel/segmentation_models.pytorch/blob/master/segmentation_models_pytorch/unet/model.py
* https://github.com/ternaus/TernausNet/blob/master/unet_models.py
* https://github.com/ternaus/TernausNetV2/blob/master/models/ternausnet2.py

It'd be nice to have a normalization layer option, ideally flexible like TernausNet. There is a worthwile reference in both of Vladimir's TernausNet impl about the transpose conv and the potential artifacts resulting from it when implemented as in the original paper.

**Comment by adamjstewart** [comment]
@fmassa What's the status of this PR? Is this something that is still wanted, or should I close it?

**Comment by adamjstewart** [comment]
It seems like this PR has been abandoned by upstream, so I'm going to close it. Feel free to reopen or steal these commits to make a new PR.
